### ISSUE TITLE
Redis node CPU usage significantly higher with Lettuce

### ISSUE THREAD

**Issue by mtheos** [lettuce-io/lettuce-core#2289]
Hi team,

We've migrated from Jedis to Lettuce and noticed a significant increase in CPU usage on our Redis node in production.
We saw CPU usage of 5-10% with Jedis and 20-40% with Lettuce.

I believe the difference is that Jedis batches requests when pipelining, whereas Lettuce is best-effort batching with auto-flushing. I say this because we also saw a significant increase in network traffic for the node when we switched to Lettuce and I did some packet capture locally and saw that in a tight loop, the largest batches with auto-flushing were ~5-10 commands (and more frequently, 1-3).

I'm now looking at disabling auto-flushing to see if doing it manually is a reasonable middle ground before connection pooling. I read somewhere that batching works best when you have a single thread and defined checkpoints, but please let me know if it's actually unsafe to share a connection between multiple threads that will be writing and flushing simultaneously.

To give an indication of load, we're doing ~25k RPS from ~80 tasks

**Comment by mp911de** [comment]
> Jedis batches requests when pipelining

To some extent. Jedis uses a buffered output stream with a default size of 8kb. In pipelining mode, Jedis doesn't flush after each command and therefore, you see a different TCP sending behavior.

The usage pattern of Jedis connection per thread where you get strong isolation against usage by other threads. Lettuce connections are intended to be shared across multiple threads and Lettuce flushes after each command.

**Comment by mtheos** [comment]
Is it an issue to share a connection that doesn't auto-flush? I don't need any synchronization when I flush manually right?

**Comment by mtheos** [comment]
Hi Mark, thanks for your reply. I've been able to determine this is the cause and by manually flushing a shared connection our CPU usage has returned to the expected level.

One correction in my original post that I misremembered, `whereas Lettuce is best-effort batching with auto-flushing` Lettuce doesn't batch at all when auto-flushing, the batches I was observing were all Redis responses, the requests were always individual.
